examples/http-protocol-versions/http3-client/main.go

```diff
@@ -0,0 +1,44 @@
+package main
```

Comment:
do we need to add an http3 client? 

afaict curl has http/3 support (at least experimentally)

Reply:
Curl does not support HTTP/3 by default. I use Fedora 34 and using `--http3` flag does not work. I had to build curl with cloudflare's quiche for that to work - in other words, a custom build is needed for this to work. So I searched for images with custom curl build. They all are old and does not seem to support IETF QUIC v1 (the standardized one), but the support is limited to older drafts. And....Using random images from Dockerhub is risky.

Does the CI environment have a curl that supports HTTP/1.1, HTTP/2 and HTTP/3? Or is there a reliable, container-based build and testing for examples as well? I see that is how builds are done. But for verifying examples it does not seem to be the case (I checked `.azure-pipelines/pipelines.yaml` and saw that verifying examples is not invoked with `ci/run_envoy_docker.sh`)

So I thought of writing a simple HTTP/3 client and package that as a docker image to use in tests (so that I don't have to push it anywhere). That is the `http3-client` folder.

Reply:
thanks for explanation

> And....Using random images from Dockerhub is risky.

yep, agreed

> Does the CI environment have a curl that supports HTTP/1.1, HTTP/2 and HTTP/3?

as these are examples its about more than just the CI env - we need something that will (mostly) work for (most) users

> So I thought of writing a simple HTTP/3 client and package that as a docker image to use in tests (so that I don't have to push it anywhere). That is the http3-client folder.

ok, understood - if we need to add it lets do so - but can we at least use docker-compose as that is how all of the other examples work

my discomfort is basically this - if we are wanting to demonstrate standards operation/compliance then it would be better not to say, "...but you will have to build a custom container/binary/etc to try it"

but if that is where the state of play is at, then yep, lets do what is required

Reply:
> my discomfort is basically this - if we are wanting to demonstrate standards operation/compliance then it would be better not to say, "...but you will have to build a custom container/binary/etc to try it"

A custom build of curl with quiche as described [here](https://github.com/curl/curl/blob/master/docs/HTTP3.md) would be equally discomforting? It is new and experimental. So HTTP/3 would be a problem from compliance perspective. Right? If compliance is a must have, then HTTP/3 has to be moved out of here into its own example. Or I can mention this specific caveat for HTTP/3?

> but can we at least use docker-compose as that is how all of the other examples work

I haven't yet experimented this one. I see some clues in grpc-bridge example.

Reply:
> If compliance is a must have, then HTTP/3 has to be moved out of here into its own example. Or I can mention this specific caveat for HTTP/3?

i wasnt implying that everything here must meet all standards all of the time

more, i just meant, this is demoing HTTP/3, its a standard, lets try and demonstrate the standard, and that custom builds confuse that a little

nevertheless, if there is not a readily available (standards-compliant) client, then i totally understand the need to build something custom in the interim.

Reply:
I've done a bunch of testing with `h3` clients - https://dev.to/moderation/experiments-with-h3-clients-envoy-52gj

None of them are trivial with `curl` being the most complicated to build. The simplest h3 client is from Istio.
```
https://github.com/istio/istio.git
cd pkg/test/echo/cmd/client
go build
```

Reply:
Thanks @moderation . Yeah. Istio uses `quic-go` package - same thing that I have used here (but simpler as it is purpose built for the test). However, I guess I can try using it and remove this client if possible. They already use echo-client in their tests anyways. So it is worth giving a shot.

Reply:
@moderation - Thinking through it again. This is a simpler and purpose-built version of what is there in Istio. So I guess it is fine to keep it this way?